html5ever/macros/match_token.rs

```diff
@@ -154,82 +155,89 @@ pub struct Tag {
     name: Option<syn::Ident>,
 }
 
-impl syn::synom::Synom for Tag {
-    named!(parse -> Self, do_parse!(
-        punct!(<) >>
-        closing: option!(punct!(/)) >>
-        name: alt!(
-            syn!(syn::Ident) => { |i| Some(i) }
-            |
-            punct!(_) => { |_| None }
-        ) >>
-        punct!(>) >>
-        (
-            Tag {
+impl Parse for Tag {
+    fn parse(input: ParseStream) -> Result<Self> {
+        input.parse::<Token![<]>()?;
+        let closing = input.parse::<Option<Token![/]>>()?;
+        let name = if let Ok(i) = input.parse::<syn::Ident>() {
+            Some(i)
+        } else {
+            input.parse::<Token![_]>()?;
+            None
+        };
+        input.parse::<Token![>]>()?;
+        Ok(Tag {
                 kind: if closing.is_some() { TagKind::EndTag } else { TagKind::StartTag },
                 name: name
-            }
-        )
-    ));
+        })
+    }
 }
 
-impl syn::synom::Synom for LHS {
-    named!(parse -> Self, do_parse!(
-        pats: alt!(
-            syn!(syn::Pat) => { |p| LHS::Pattern(p) }
-            |
-            many0!(syn!(Tag)) => { |t| LHS::Tags(t) }
-        ) >>
-        (
-            pats
-        )
-    ));
+impl Parse for LHS {
+    fn parse(input: ParseStream) -> Result<Self> {
+        if let Ok(p) = input.fork().parse::<syn::Pat>() {
+            input.parse::<syn::Pat>().unwrap();
+            Ok(LHS::Pattern(p))
+        } else {
+            let mut tags = vec![];
+            while !input.peek(Token![=>]) {
+                tags.push(input.parse()?);
+            }
+            Ok(LHS::Tags(tags))
+        }
+    }
 }
 
-impl syn::synom::Synom for MatchTokenArm {
-    named!(parse -> Self, do_parse!(
-        binding: option!(
-            do_parse!(
-                name: syn!(syn::Ident) >>
-                punct!(@) >>
-                (
-                    name
-                )
-            )
-        ) >>
-        lhs: syn!(LHS) >>
-        punct!(=>) >>
-        rhs: do_parse!(
-                expr: alt!(
-                    expr_nosemi => { |e| RHS::Expression(e) }
-                    |
-                    syn!(syn::Expr) => { |e| RHS::Expression(e) }
-                    |
-                    keyword!(else) => { |_| RHS::Else }
-                ) >>
-                option!(punct!(,)) >>
-                (expr)
-        ) >>
-        (
-            MatchTokenArm {
-               binding,
-               lhs,
-               rhs,
-            }
-        )
-    ));
+impl Parse for MatchTokenArm {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let binding = if input.peek2(Token![@]) {
+            let binding = input.parse::<syn::Ident>()?;
+            input.parse::<Token![@]>()?;
+            Some(binding)
+        } else {
+            None
+        };
+        let lhs = input.parse::<LHS>()?;
+        input.parse::<Token![=>]>()?;
+        let rhs = if let Ok(expr) = input.fork().parse() {
```

Comment:
This is probably not doing what you intended. This is bleeding over into the next arm and relying on the fact that `{ /* ... */ } < TAG > =>` is not a valid expression -- because `=>` is not a valid beginning of expression. You should parse each arm without parsing the LHS of the next arm as part of the previous arm's expression.